src/main.ts

```diff
@@ -65,6 +66,105 @@ const DEFAULT_SETTINGS: ReadwisePluginSettings = {
   reimportShowConfirmation: true,
 };
 
+class AdvancedModal extends Modal {
+  plugin: ReadwisePlugin;
+
+  constructor(app: App, plugin: ReadwisePlugin) {
+    super(app);
+    this.plugin = plugin;
+  }
+
+  /** helper to extract all book IDs from frontmatter in the readwiseDir ("base folder"),
+  perfectly matching them to the document they came from */
+  async extractBookIDs() {
+    console.log('Readwise Official plugin: extracting book IDs from frontmatter...');
+    const files = this.plugin.app.vault.getMarkdownFiles();
+    const bookIDs: { [bookID: string]: string } = {};
+    for (const file of files) {
+      if (file.path.startsWith(this.plugin.settings.readwiseDir)) {
+        const cache = this.app.metadataCache.getFileCache(file);
+        // skip if there's no cache
+        if (!cache) continue;
+
+        const frontmatter = cache.frontmatter;
+        // skip if there's no frontmatter
+        if (!frontmatter) continue;
+
+        const bookID = frontmatter[this.plugin.settings.frontmatterBookIdKey];
+        if (bookID) bookIDs[file.path] = bookID;
+      }
+    }
+
+    return bookIDs;
+  }
+
+  onOpen() {
+    const { titleEl, contentEl } = this;
+    titleEl.setText('Advanced settings');
+
+    // add new setting that allows text input
+    new Setting(contentEl)
+      .setName('Book ID frontmatter key')
+      .setDesc('The frontmatter key used to store the Readwise book_id. If set, will be used to identify the export for syncs.')
+      .addText(text => text
+        .setPlaceholder('readwise-book-id')
+        .setValue(this.plugin.settings.frontmatterBookIdKey)
+        .onChange(async value => {
+          console.log('Readwise Official plugin: setting frontmatter book ID key to', value);
+          this.plugin.settings.frontmatterBookIdKey = value;
+          this.plugin.saveSettings();
+
+          // test the new setting
+          const bookIDs = await this.extractBookIDs();
+          console.log('Readwise Official plugin: found files with book IDs in frontmatter', bookIDs);
+        }));
+
+    // button to add populate the frontmatterBookIdKey setting
+    // with whatever's saved in booksIDsMap
+    new Setting(contentEl)
+      .setName('Populate frontmatter key')
+      .setDesc('If the book ID frontmatter key is configured above, this will populate the frontmatter key setting with the first key found in the booksIDsMap. This will not overwrite existing keys.')
+      .addButton(button => button
+        .setButtonText('Populate')
+        .onClick(async () => {

```

Comment:
Note that i wanted to do something like this...
```suggestion
        .setButtonText('Populate')
        .setDisabled(!this.plugin.settings.frontmatterBookIdKey)
        .onClick(async () => {

```
... but it's trickier than expected - I'm not sure how to update the button between enabled and disabled state when the `frontmatterBookIdKey` text input changes.